cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll

```diff
@@ -160,37 +157,19 @@ private predicate boundFlowStepSsa(
  * - `upper = true`  : `op <= bound + delta`
  * - `upper = false` : `op >= bound + delta`
  */
-private IRGuardCondition boundFlowCond(NonPhiOperand op, NonPhiOperand bound, int delta, boolean upper,
+private IRGuardCondition boundFlowCond(ValueNumber vn, NonPhiOperand bound, int delta, boolean upper,
   boolean testIsTrue)
 {
-  exists(Operand compared |
-    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
-    result.controls(op.getInstruction().getBlock(), testIsTrue) and
-    valueNumber(compared.getDefinitionInstruction()) = valueNumber(op.getDefinitionInstruction())
-  )
-  // TODO: strengthening through modulus library
-}
-
-/**
- * Gets a condition that tests whether `op` is bounded by `bound + delta`.
- * 
- * - `upper = true`  : `op <= bound + delta`
- * - `upper = false` : `op >= bound + delta`
- */
-private IRGuardCondition boundFlowCondPhi(PhiOperand op, NonPhiOperand bound, int delta, boolean upper,
-  boolean testIsTrue)
-{
-  exists(Operand compared |
-    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
-    result.controlsEdgeDirectly(op.getPredecessorBlock().getLastInstruction(), op.getInstruction().getBlock(), testIsTrue) and
-    valueNumber(compared.getDefinitionInstruction()) = valueNumber (op.getDefinitionInstruction())
+  exists(int d |
+    result.comparesLt(vn.getAUse(), bound, d, upper, testIsTrue) and
+    // strengthen from x < y to x <= y-1
+    if upper = true
```

Comment:
This strengthening appears wrong.  There are at least four cases to be handled for each of the true-false combinations of `upper` and `testIsTrue`.  I'd think that in two of those cases we'd have `delta = d`, one case with `delta = d-1`, and one case with `delta = d+1`.

Reply:
The C++ guards library's `comparelt` provides the same semantics as `boundFlowCond` but with a `<` rather than a `<=` when `upper = true`. I've added a test to demonstrate.

Reply:
Ah, right.